File: books/data-structures/day2_array/best-time-to-buy-and-sell-stock.cc
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = (int)prices.size(), ans = 0;
        for (int i = 0; i < n; ++i) {
            for (int j = i+1; j < n; ++j) {
                ans = max(ans, prices[j] - prices[i]);
            }
        }
        return ans;
    }

    int maxProfit2(vector<int>& prices) {
        int inf = 1e9;
        int min_price = inf, max_profit = 0;

        for (int price: prices) {
            max_profit = max(max_profit, price - min_price);
            min_price = min(price, min_price);
        }
        return max_profit;
    }
};

File: books/data-structures/day2_array/best-time-to-buy-and-sell-stock.rs

#[warn(dead_code)]
struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min = if prices.len() > 0 { prices[0] } else { 0 };
        let mut max = 0;

        for i in 1..prices.len() {
            max = max.max(prices[i] - min);
            if prices[i] < min { min = prices[i]; }
        }
        max
    }
}

fn main() {

}